apps/site/lib/site_web/controllers/schedule/line/diagram_format.ex

```diff
@@ -66,19 +66,35 @@ defmodule SiteWeb.ScheduleController.Line.DiagramFormat do
 
   defp has_active_diversion?(_, _), do: false
 
-  defp chunk_by_diversions({stop, index}, chunk, date) do
+  @spec effects_for_stop(line_diagram_stop) :: [{String.t(), Alerts.Alert.effect()}]
+  defp effects_for_stop(%{alerts: alerts}) do
+    Enum.map(alerts, &{&1.id, &1.effect})
+  end
+
+  @spec chunk_by_diversions(line_diagram_stop(), [line_diagram_stop()], DateTime.t()) ::
+  {:cont, any(), [line_diagram_stop()]}
+  | {:cont, [line_diagram_stop()]}
+  | {:halt, [line_diagram_stop()]}
+  defp chunk_by_diversions(stop, chunk, date) do
     if stop_has_diversion_now?(stop, date) do
-      stop_effects = Enum.map(stop.alerts, &{&1.id, &1.effect})
+      stop_fx = effects_for_stop(stop)
 
       case chunk do
         [] ->
-          {:cont, [{index, stop_effects}]}
+          {:cont, [stop]}
 
-        [{_, last_effects} | _] ->
-          if stop_effects == last_effects do
-            {:cont, Enum.reverse([{index, stop_effects} | chunk]), []}
+        chunk ->
+          last_stop = List.last(chunk)
+          if stop_fx == effects_for_stop(last_stop) do
+            {:cont, chunk ++ [stop]}
           else
-            {:cont, [{index, stop_effects}]}
+            case stop_fx do
```

Comment:
I generally think of multiple types of conditionals (e.g. a `case` within an `if`) to be a red flag, potentially a sign that the conditionals should be combined or the function is responsible for too many things and could use refactoring. This is now 4 layers of nested conditionals deep. Are there ways we could consider refactoring out sub-functions that might each handle a single thing?